Improve helm for Version 3 (WIP) (#1340)
api-platform/docs: deployment/kubernetes.md

# Deploying to a Kubernetes Cluster

[Kubernetes](https://kubernetes.io/) has become the most popular way to deploy, run and manage containers in production.
[Google Cloud Platform](https://cloud.google.com/kubernetes-engine/), [Microsoft Azure](https://azure.microsoft.com/en-us/services/container-service/kubernetes/)
and [Amazon Web Services](https://aws.amazon.com/eks/) and many more companies provide managed Kubernetes environment.

[The official API Platform distribution](../distribution/index.md) contains a built-in [Helm](https://helm.sh/) (the k8s
package manager) chart to deploy in a wink on any of these platforms.

> This guide is based on Helm 3 and the caddy/php/pwa/postgres Stack introduced with API Platform 2.6

Reviewer: ```suggestion
This guide is based on Helm 3.
```

## Preparing Your Cluster and Your Local Machine

1. Create a Kubernetes cluster on your preferred Cloud provider or install Kubernetes locally on your servers by example with [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)
2. Install [Helm](https://helm.sh/) (preferred version >= 3) `locally` and on your `cluster` following their [documentation](https://helm.sh/docs/intro/install/)
3. Be sure to be connected to the right Kubernetes container
   `kubectl config view` [Details](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/)
   e.g. for Google Cloud running: `gcloud config get-value core/project`
4. Update the Helm repo on your local machine: `helm repo update`
   If you have a fresh installation of helm, you do not have any repositories yet.

> Working-Dir: Your local installation of api-platform. Default /api-platform/

Reviewer: Why this line?

Author: I would recommend to remove the Point 4. in the list. Because it does nothing useful on an fresh installation. Maybe this was added for helm2?

- Working-Dir: Your local installation of api-platform. Default /api-platform/
I just want to be sure that the path for the following commands are correct. I am not sure if this is clear to everyone.

Reviewer: Maybe a complete sentence would be better then?

## Creating and Publishing the Docker Images

### Example with the [Google Container Registry](https://cloud.google.com/container-registry/)

> #### 1. Build the PHP and Caddy Docker images and tag them

Reviewer: Why not using the list?

    docker build -t gcr.io/test-api-platform/php -t gcr.io/test-api-platform/php:latest api --target api_platform_php
    docker build -t gcr.io/test-api-platform/caddy -t gcr.io/test-api-platform/caddy:latest api --target api_platform_caddy
    docker build -t gcr.io/test-api-platform/pwa -t gcr.io/test-api-platform/pwa:latest pwa --target api_platform_pwa_prod

#### 2. Push your images to your Docker registry

    gcloud auth configure-docker
    docker push gcr.io/test-api-platform/php
    docker push gcr.io/test-api-platform/caddy
    docker push gcr.io/test-api-platform/pwa

## Deploying with Helm 3

> ### 1. Check the Helm-Version

Reviewer: ```suggestion
### 1. Check the Helm version
```

    helm version

> If you are using version 2.x follow this [guid to migrating Helm to v3](https://helm.sh/docs/topics/v2_v3_migration/#helm)

Reviewer: ```suggestion
If you are using version 2.x follow this [guide to migrate Helm to v3](https://helm.sh/docs/topics/v2_v3_migration/#helm)
```

### 2. Firstly you need to update helm dependencies by running

    helm dependency update ./helm/api-platform

> ### 3. Optional: If you made changes to the Helm-chart check if format is correct

Reviewer: ```suggestion
### 3. Optional: If you made changes to the Helm chart, check if its format is correct
```

    helm lint ./helm/api-platform

### 4. Deploy your API to the container

    helm install api-platform ./helm/api-platform --namespace=default \
        --set "php.image.repository=gcr.io/test-api-platform/php" \
        --set php.image.tag=latest \
        --set "caddy.image.repository=gcr.io/test-api-platform/caddy" \
        --set caddy.image.tag=latest \
        --set "pwa.image.repository=gcr.io/test-api-platform/pwa" \
        --set pwa.image.tag=latest \
        --set php.appSecret='!ChangeMe!' \
        --set postgresql.postgresqlPassword='!ChangeMe!' \
        --set postgresql.persistence.enabled=true \
        --set corsAllowOrigin='^https?://[a-z\]*\.mywebsite.com$'

You can add the parameter `--dry-run` to check upfront if anything is correct.

> Replace the values with the image-parameters from the stage above.

Reviewer: ```suggestion
Replace the values with the image parameters from the stage above.
```

# Deploying to a Kubernetes Cluster

[Kubernetes](https://kubernetes.io/) has become the most popular way to deploy, run and manage containers in production.
[Google Cloud Platform](https://cloud.google.com/kubernetes-engine/), [Microsoft Azure](https://azure.microsoft.com/en-us/services/container-service/kubernetes/)
and [Amazon Web Services](https://aws.amazon.com/eks/) and many more companies provide managed Kubernetes environment.

[The official API Platform distribution](../distribution/index.md) contains a built-in [Helm](https://helm.sh/) (the k8s
package manager) chart to deploy in a wink on any of these platforms.

This guide is based on Helm 3.

## Preparing Your Cluster and Your Local Machine

> 1. Create a Kubernetes cluster on your preferred Cloud provider or install Kubernetes locally on your servers by example with [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)

Reviewer 2: ```suggestion
1. Create a Kubernetes cluster on your preferred Cloud provider or install Kubernetes locally on your server, for example with [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)
```

> 2. Install [Helm](https://helm.sh/) (preferred version >= 3) `locally` and on your `cluster` following their [documentation](https://helm.sh/docs/intro/install/)

Reviewer 2: ```suggestion
2. Install [Helm 3](https://helm.sh/) `locally` and on your `cluster` following their [documentation](https://helm.sh/docs/intro/install/)
```

    docker push gcr.io/test-api-platform/caddy
    docker push gcr.io/test-api-platform/pwa

Optional push the version images

    docker push gcr.io/test-api-platform/php:0.1.0 
    docker push gcr.io/test-api-platform/caddy:0.1.0 
    docker push gcr.io/test-api-platform/pwa:0.1.0 

The result should look similar to these images.

![Example of Google Images - Overview](images/google-image-overview.png)
![Example of Google Caddy Image - Details](images/google-image-caddy-details.png)

## Deploying with Helm 3

### 1. Check the Helm version

    helm version

If you are using version 2.x follow this [guide to migrate Helm to v3](https://helm.sh/docs/topics/v2_v3_migration/#helm)

### 2. Firstly you need to update helm dependencies by running

    helm dependency update ./helm/api-platform

This will create a folder /charts/ and add all dependencies there.
Actual this is [bitnami/postgresql](https://bitnami.com/stack/postgresql/helm), a file postgresql-[VERSION].tgz is created.

### 3. Optional: If you made changes to the Helm chart, check if its format is correct

    helm lint ./helm/api-platform

### 4. Deploy your API to the container

>     helm install api-platform ./helm/api-platform --namespace=default \

Reviewer 2: FYI, you can find a complete deploy command on the demo project: https://github.com/api-platform/demo/blob/main/.github/workflows/deploy.yml

Author: Thx for the info. I will add it to the Docu, because i think that will help others too

    kubectl --namespace=default exec -it $CADDY_PHP_POD -c api-platform-php -- bin/console

## Caution for system architecture

If the pods do not run, and you get the following error from google kubernetes engine logs,
there is probably a problem with the system architecture.
`standard_init_linux.go:211: exec user process caused "exec format error`
Build the images with the same system architecture as the cluster runs.
Example: Building with Mac M1 with arm64 leads to problems. Most cluster will run with x86_64.
Solution: <https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/>

## Updates

There are 2 main upgrade strategies.

### 1. Always version your images (recommended)

Change the version in the attribut "appVersion" in Chart.yaml and tag the images.
You can upgrade with the same command from the installation and pass all parameters.
Just replace `helm install ...` with `helm upgrade ...`

### 2. Use :latest tags

Infos about [best practices for tagging images for kubernetes](https://kubernetes.io/docs/concepts/containers/images/)

> You habe to use the imagePullPolicy=Always see the last 3 parameters.

Reviewer 2: ```suggestion
You have to use the imagePullPolicy=Always see the last 3 parameters.
```

    docker build -t gcr.io/test-api-platform/caddy:0.1.0 -t gcr.io/test-api-platform/caddy:latest api --target api_platform_caddy
    docker build -t gcr.io/test-api-platform/pwa:0.1.0 -t gcr.io/test-api-platform/pwa:latest pwa --target api_platform_pwa_prod

#### 2. Push your images to your Docker registry

    gcloud auth configure-docker
    docker push gcr.io/test-api-platform/php
    docker push gcr.io/test-api-platform/caddy
    docker push gcr.io/test-api-platform/pwa

Optional push the version images

    docker push gcr.io/test-api-platform/php:0.1.0 
    docker push gcr.io/test-api-platform/caddy:0.1.0 
    docker push gcr.io/test-api-platform/pwa:0.1.0 

The result should look similar to these images.

![Example of Google Images - Overview](images/google-image-overview.png)
![Example of Google Caddy Image - Details](images/google-image-caddy-details.png)

## Deploying with Helm 3

### 1. Check the Helm version

    helm version

If you are using version 2.x follow this [guide to migrate Helm to v3](https://helm.sh/docs/topics/v2_v3_migration/#helm)

### 2. Firstly you need to update helm dependencies by running

>     helm dependency update ./helm/api-platform

Reviewer 2: Does it work on your install? On api-platform/demo, I had to do the following;
```
helm repo add bitnami https://charts.bitnami.com/bitnami/
helm repo add stable https://charts.helm.sh/stable/
helm dependency build ./helm/api-platform
```

> This will create a folder /charts/ and add all dependencies there.

Reviewer 2: ```suggestion
This will create a folder helm/api-platform/charts/ and add all dependencies there.
```

> Actual this is [bitnami/postgresql](https://bitnami.com/stack/postgresql/helm), a file postgresql-[VERSION].tgz is created.

Reviewer 2: I don't understand the first part of this sentence...

### 3. Optional: If you made changes to the Helm chart, check if its format is correct

    helm lint ./helm/api-platform

### 4. Deploy your API to the container

>     helm upgrade api-platform ./helm/api-platform --namespace=default \

Reviewer 2: I recommend naming the release differently than `api-platform` to prevent the k8s services be named `api-platform-apl-piatform`:
```suggestion
    helm upgrade main ./helm/api-platform --namespace=default --create-namespace --wait \
```

>         --install \

Reviewer 2: What about the ingress and mercure?

        --set "php.image.repository=gcr.io/test-api-platform/php" \
        --set php.image.tag=latest \
        --set "caddy.image.repository=gcr.io/test-api-platform/caddy" \
        --set caddy.image.tag=latest \
        --set "pwa.image.repository=gcr.io/test-api-platform/pwa" \
        --set pwa.image.tag=latest \
        --set php.appSecret='!ChangeMe!' \
        --set postgresql.postgresqlPassword='!ChangeMe!' \
        --set postgresql.persistence.enabled=true \

>         --set "corsAllowOrigin=^https?://[a-z\]*\.mywebsite.com$"

Reviewer 2: ```suggestion
        --set "corsAllowOrigin=^https?:\/\/[a-z]*\.mywebsite.com$"
```

> The " are necessary for Windows. Use ^ on Windows instead of \ to split commands into multiple lines.

Reviewer 2: ```suggestion
The `"` are necessary for Windows. Use ^ on Windows instead of \ to split commands into multiple lines.
```

You can add the parameter `--dry-run` to check upfront if anything is correct.
Replace the values with the image parameters from the stage above.
The parameter `php.appSecret` is the `AppSecret` from ./.env
Fill the rest of the values with the correct settings.
For available options see /helm/api-platform/values.yaml.
If you want a test deploy you can set corsAllowOrigin='*'

After a successful installation, there is a message at the end.
You can copy these commands and execute them to set a port-forwarding and
get access on your local machine to the deploy. See image below.

![Deploy Result](images/deploy-result.png)

If you prefer to use a managed DBMS like [Heroku Postgres](https://www.heroku.com/postgres) or
[Google Cloud SQL](https://cloud.google.com/sql/docs/postgres/) (recommended):

    helm upgrade api-platform ./helm/api-platform \
        # ...
        --set postgresql.enabled=false \
        --set postgresql.url=pgsql://username:password@host/database?serverVersion=13

> Finally, build the `pwa` (client and admin) JavaScript apps and [deploy them on a static

Reviewer 2: We don't use static version built of the PWA on api-platform, now we're using [Next.js static generation](https://nextjs.org/docs/basic-features/data-fetching). You can have a look at the [demo PWA](https://github.com/api-platform/demo/tree/main/pwa) as an example.

But in a sense, you're still right as building the PWA and deploying it on a static server is still possible.

Build the images with the same system architecture as the cluster runs.
Example: Building with Mac M1 with arm64 leads to problems. Most cluster will run with x86_64.
Solution: <https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/>

## Updates

There are 2 main upgrade strategies.

### 1. Always version your images (recommended)

Change the version in the attribut "appVersion" in Chart.yaml and tag the images with this version.
You can upgrade with the same command from the installation and pass all parameters.

### 2. Use :latest tags

Infos about [best practices for tagging images for kubernetes](https://kubernetes.io/docs/concepts/containers/images/)
You have to use the *.image.pullPolicy=Always see the last 3 parameters.

    helm upgrade api-platform ./helm/api-platform --namespace=default \
    --set "php.image.repository=gcr.io/test-api-platform/php" \
    --set php.image.tag=latest \
    --set "caddy.image.repository=gcr.io/test-api-platform/caddy" \
    --set caddy.image.tag=latest \
    --set "pwa.image.repository=gcr.io/test-api-platform/pwa" \
    --set pwa.image.tag=latest \
    --set php.appSecret='!ChangeMe!' \

>     --set postgresql.postgresqlPassword='!ChangeMe!' \

Reviewer 2: While updating an existing database, this might cause issues. I don't recommend to change the database password on update. That's why I had to do [the following on api-platform/demo](https://github.com/api-platform/demo/blob/main/.github/workflows/deploy.yml#L187):
```
if ! kubectl get namespace $GITHUB_REF_SLUG > /dev/null 2>&1; then
    // install
else
    // update
fi
```
